src/createSlice.ts

```diff
@@ -0,0 +1,115 @@
+import { Action, AnyAction, ActionCreator, Reducer } from 'redux'
+import { createAction, PayloadAction } from './createAction'
+import { createReducer, CaseReducersMapObject } from './createReducer'
+import { createSliceSelector, createSelectorName } from './sliceSelector'
+
+export interface Slice<
+  S = any,
+  A extends Action = AnyAction,
+  AT extends string = string
+> {
+  /**
+   * The slice name.
+   */
+  slice: string
+
+  /**
+   * The slice's reducer.
+   */
+  reducer: Reducer<S, A>
+
+  /**
+   * Action creators for the types of actions that are handled by the slice
+   * reducer.
+   */
+  actions: { [type in AT]: ActionCreator<A> }
```

Comment:
After you fix the type of `PayloadAction`:

```ts
type ActionTypes<A extends Action> = A extends { type: infer T } ? T : never
type ActionWithType<A extends Action, T extends string> = Extract<A, { type: T }>
```

```ts
actions: { [T in ActionTypes<A>]: ActionCreator<ActionWithType<A, T>>
```

The `AT` generic argument is unneeded / redundant.

Reply:
This didn't work. TypeScript inferred the key type of `slice.actions` as `string` in one of the typing tests, so it was possible to write `slice.actions.nonExistingAction`.